cmds/webboot/webboot.go

```diff
@@ -0,0 +1,190 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"regexp"
+
+	ui "github.com/gizak/termui/v3"
+	"github.com/u-root/webboot/pkg/menu"
+)
+
+const (
+	downloadByLinkLabel = "Download by link"
+	tcURL               = "http://tinycorelinux.net/10.x/x86_64/release/TinyCorePure64-10.1.iso"
+	wbtcpURL            = "https://github.com/u-root/webboot-distro/raw/master/iso/tinycore/10.x/x86_64/release/TinyCorePure64.iso"
+)
+
+// bookmark record the list of iso webboot allow user to download
+var bookmark = []*BookMarkISO{
+	&BookMarkISO{
+		isDefault: false,
+		url:       tcURL,
+		label:     "Download Tinycore v10.1",
+		isoName:   "TinyCorePure64-10.1.iso",
+	},
+	&BookMarkISO{
+		isDefault: false,
+		url:       wbtcpURL,
+		label:     "Webboot Tinycorepure",
+		isoName:   "TinyCorePure64.iso",
+	},
+}
+
+// BookMarkISO contains information of the iso user want to download
+// expected to implement Entry interface
+type BookMarkISO struct {
+	isDefault bool
+	url       string
+	label     string
+	isoName   string
+}
+
+// Label is the string this iso displays in the menu page.
+func (b *BookMarkISO) Label() string {
+	return b.label
+}
+
+// IsDefault is for mark whether this iso is a default choice.
+func (b *BookMarkISO) IsDefault() bool {
+	return b.isDefault
+}
+
+// Exec performs the following process after the entry is chosen
+func (b *BookMarkISO) Exec(_ <-chan ui.Event) error {
+	link := b.url
+	filename := b.isoName
+	fPath := filepath.Join("/tmp", filename)
+	if err := download(link, fPath); err != nil {
+		return err
+	}
+	// todo: boot the iso
+	return nil
+}
+
+// CachedISO contains information of the iso cached in the memory
+// expected to implement Entry interface
+type CachedISO struct {
+	isDefault bool
+	label     string
+	path      string
+	isoName   string
+	// todo: information parsed from config file
+}
+
+// DownloadByBookmark is to implement "Download by bookmark" option in the menu
+type DownloadByBookmark struct{}
+
+// Label is the string this iso displays in the menu page.
+func (d *DownloadByBookmark) Label() string {
+	return "Download by link"
+}
+
+// IsDefault is for mark whether this iso is a default choice.
+// assume that DownloadByLink will not be a default option.
+func (d *DownloadByBookmark) IsDefault() bool {
+	return false
+}
+
+// Exec performs the following process after the entry is chosen
+func (d *DownloadByBookmark) Exec(uiEvents <-chan ui.Event) error {
+	entries := []menu.Entry{}
+	for _, e := range bookmark {
+		entries = append(entries, e)
+	}
+
+	chosen, err := menu.DisplayMenu("Bookmarks", "Input your choice", entries, uiEvents)
+	if err != nil {
+		return err
+	}
+
+	return chosen.Exec(uiEvents)
```

Comment:
Do you need to do this, won't DisplayMenu call the Exec function for the chosen entry?

Reply:
no it won't currently. you think it would be better to run Exec in the DisplayMenu?

Reply:
I thought that was already happening, sorry I missed that change - hm I think one of the advantages of having Exec is that DisplayMenu can run it, otherwise why do we need that as part of the Entry interface? I guess its not necessary, but I'd do it from DisplayMenu for convenience. If anyone wants to wait until the function returns, they could always pass an empty Exec function?